### ISSUE TITLE
Possible problem with data synchronization on packet boundary.  (ros ticket #3890)

### ISSUE THREAD

**Issue by ahendrix** [PR2/pr2_ethercat_drivers#18]
Problem with data synchronization on packet boundary.

In a recent version of pr2_etherCAT, verification of checksum values on status data was added.   Since that time, the r_forearm_roll MCB on PRL has reported checksum errors multiple times over the course of a week.  An estimate of the MTBF is 30 minutes.   Initially it was assumed that there was a bad solder joint on that particular MCB.  After replacing the problem board in PRL, we are still seeing the same issues with issue.  Also there was a report today of a checksum error on the r_forearm_roll motor on PRH. 

I'm now under the assumption that there is some sort of race condition that only effects that particular board.  I've looked how the process data is mapped on PRL and have found that the status data of device 30 (r_forearm_roll) is being split across two packets.

Debuging output from pr2_ethercat showing addresses that each devices status and command data is mapped to (mapping start at address 0x10000). 
{{{ 
[ INFO] 1266471784.581650000: device 2, command  10000 = 0x10000+0
[ INFO] 1266471784.581825000: device 2, status   10008 = 0x10000+8
[ INFO] 1266471784.625008000: device 3, command  10034 = 0x10000+52
[ INFO] 1266471784.625054000: device 3, status   1003C = 0x10000+60
[ INFO] 1266471784.668233000: device 4, command  10068 = 0x10000+104
[ INFO] 1266471784.668278000: device 4, status   10070 = 0x10000+112
[ INFO] 1266471784.711455000: device 5, command  1009C = 0x10000+156
[ INFO] 1266471784.711501000: device 5, status   100A4 = 0x10000+164
[ INFO] 1266471784.754678000: device 6, command  100D0 = 0x10000+208
[ INFO] 1266471784.754727000: device 6, status   100D8 = 0x10000+216
[ INFO] 1266471784.797903000: device 7, command  10104 = 0x10000+260
[ INFO] 1266471784.797948000: device 7, status   1010C = 0x10000+268
[ INFO] 1266471784.841132000: device 8, command  10138 = 0x10000+312
[ INFO] 1266471784.841182000: device 8, status   10140 = 0x10000+320
[ INFO] 1266471784.884350000: device 9, command  1016C = 0x10000+364
[ INFO] 1266471784.884398000: device 9, status   10174 = 0x10000+372
[ INFO] 1266471784.927575000: device 10, command  101A0 = 0x10000+416
[ INFO] 1266471784.927621000: device 10, status   101A8 = 0x10000+424
[ INFO] 1266471784.970801000: device 11, command  101D4 = 0x10000+468
[ INFO] 1266471784.970847000: device 11, status   101DC = 0x10000+476
[ INFO] 1266471785.014023000: device 12, command  10208 = 0x10000+520
[ INFO] 1266471785.014105000: device 12, status   10210 = 0x10000+528
[ INFO] 1266471785.057249000: device 13, command  1023C = 0x10000+572
[ INFO] 1266471785.057296000: device 13, status   10244 = 0x10000+580
[ INFO] 1266471785.100472000: device 14, command  10270 = 0x10000+624
[ INFO] 1266471785.100519000: device 14, status   10278 = 0x10000+632
[ INFO] 1266471785.227145000: device 17, command  102A4 = 0x10000+676
[ INFO] 1266471785.227194000: device 17, status   102AC = 0x10000+684
[ INFO] 1266471785.270370000: device 18, command  102D8 = 0x10000+728
[ INFO] 1266471785.270417000: device 18, status   102E0 = 0x10000+736
[ INFO] 1266471785.313593000: device 19, command  1030C = 0x10000+780
[ INFO] 1266471785.313642000: device 19, status   10314 = 0x10000+788
[ INFO] 1266471785.356820000: device 20, command  10340 = 0x10000+832
[ INFO] 1266471785.356876000: device 20, status   10348 = 0x10000+840
[ INFO] 1266471785.400042000: device 21, command  10374 = 0x10000+884
[ INFO] 1266471785.400089000: device 21, status   1037C = 0x10000+892
[ INFO] 1266471785.443256000: device 22, command  103A8 = 0x10000+936
[ INFO] 1266471785.443298000: device 22, status   103B0 = 0x10000+944
[ INFO] 1266471785.486491000: device 23, command  103DC = 0x10000+988
[ INFO] 1266471785.486537000: device 23, status   103E4 = 0x10000+996
[ INFO] 1266471785.529773000: device 24, command  10410 = 0x10000+1040
[ INFO] 1266471785.529816000: device 24, status   10418 = 0x10000+1048
[ INFO] 1266471785.529850000: device 24, pressure 10455 = 0x10000+1109
[ INFO] 1266471785.573687000: device 25, command  104B3 = 0x10000+1203
[ INFO] 1266471785.573735000: device 25, status   104BB = 0x10000+1211
[ INFO] 1266471785.616914000: device 26, command  104E7 = 0x10000+1255
[ INFO] 1266471785.616960000: device 26, status   104EF = 0x10000+1263
[ INFO] 1266471785.660137000: device 27, command  1051B = 0x10000+1307
[ INFO] 1266471785.660185000: device 27, status   10523 = 0x10000+1315
[ INFO] 1266471785.703365000: device 28, command  1054F = 0x10000+1359
[ INFO] 1266471785.703411000: device 28, status   10557 = 0x10000+1367
[ INFO] 1266471785.746585000: device 29, command  10583 = 0x10000+1411
[ INFO] 1266471785.746632000: device 29, status   1058B = 0x10000+1419
[ INFO] 1266471785.789808000: device 30, command  105B7 = 0x10000+1463
[ INFO] 1266471785.789865000: device 30, status   105BF = 0x10000+1471
[ INFO] 1266471785.833033000: device 31, command  105EB = 0x10000+1515
[ INFO] 1266471785.833079000: device 31, status   105F3 = 0x10000+1523
[ INFO] 1266471785.876257000: device 32, command  1061F = 0x10000+1567
[ INFO] 1266471785.876304000: device 32, status   10627 = 0x10000+1575
[ INFO] 1266471785.919482000: device 33, command  10653 = 0x10000+1619
[ INFO] 1266471785.919528000: device 33, status   1065B = 0x10000+1627
[ INFO] 1266471785.919560000: device 33, pressure 10698 = 0x10000+1688
[ INFO] 1266471785.963204000: device 34, command  106F6 = 0x10000+1782
[ INFO] 1266471785.963252000: device 34, status   106FE = 0x10000+1790
[ INFO] 1266471786.006476000: device 35, command  1072A = 0x10000+1834
[ INFO] 1266471786.006535000: device 35, status   1073A = 0x10000+1850
[ INFO] 1266471786.049655000: device 36, command  10766 = 0x10000+1894
[ INFO] 1266471786.049701000: device 36, status   1076E = 0x10000+1902
}}}

The amount of data we allow in one packet is 1486 bytes (specifed in eml / ethercat_process_data.cxx).  
The first 15 bytes of r_forearm_roll data is in the first packet (1486-1471 = 15).   The last 29 bytes in the second packet (1515-1486 = 29).

Splitting process data across a packet boundary should NOT cause issues, because the EtherCAT SyncManager hardware should provide atomicity to both command and status buffer access.  When the first byte of status buffer is read, the SyncManager will prevents changes to the contents of the buffer until the last byte of the buffer is read.

trac data:
- Owner: **dking**
- Reporter: **dking**
- Reported at: **Wed Feb 17 22:46:14 2010**
- URL: https://code.ros.org/trac/wg-ros-pkg/ticket/3890

**Comment by ahendrix** [comment]
[dking] Checksum error with original board on PRL.
{{{
Component: EtherCAT Device (r_forearm_roll_motor)
Hardware ID: 68-05005-01390

Level: Error
Message: Checksum error on status data

Configuration: good
Name: r_forearm_roll_motor
Position: 30

Product code: WG05 (6805005) Firmware Revision 1.20, PCB Revision F.02
Robot: PR2
Motor: Maxon 310009
Serial Number: 68-05005-01390

Nominal Current Scale: 0.000500
Nominal Voltage Scale: 0.006250
HW Max Current: 3.250000
SW Max Current: 1.720000
Speed Constant: 178.000000
Resistance: 2.520000
Motor Torque Constant: 0.053800
Pulses Per Revolution: 1200
Encoder Reduction: -1.000000
Safety Disable Status: ENABLED (00)
Safety Disable Status Hold: ENABLED (00)
Safety Disable Count: 66
Undervoltage Count: 42

Over Current Count: 0
Board Over Temp Count: 0
Bridge Over Temp Count: 0
Operate Disable Count: 1
Watchdog Disable Count: 25
MBX Command IRQ Count: 52

PDI Timeout Error Count: 0
PDI Checksum Error Count: 0
Supply Current: 0.152796
Configured Offset A: 0.008000
Configured Offset B: 0.036500

Mailbox Write Errors: 0
Mailbox Read Errors: 0
Mailbox Retries: 0
Mailbox Retry Errors: 0
Watchdog Limit: 100ms
Mode: OFF
Digital out: 254
Programmed pwm value: 0
Programmed current: 0.000000
Measured current: -0.008000
Timestamp: 2538700428
Encoder count: 341788
Encoder index pos: 341788
Num encoder_errors: 0
Encoder status: 12
Calibration reading: 15
Last calibration rising edge: 341725
Last calibration falling edge: 341696
Board temperature: 39.625000
Max board temperature: 40.750000
Bridge temperature: 42.375000
Max bridge temperature: 46.000000
Supply voltage: 70.518753
Motor voltage: -0.093750

Current Loop Kp: 16
Current Loop Ki: 4
Motor Voltage Error %: 5.498262
Max Motor Voltage Error %: 55.597393
Filtered Voltage Error %: 5.603648
Max Filtered Voltage Error %: 52.664371
Current Error: 0.000004
Max Current Error: 0.046984
Filtered Current Error: 0.000080
Max Filtered Current Error: 0.006501
Motor Resistance Estimate: 6.520463
Packet count: 7778

Drops: 3
Consecutive Drops: 0
Max Consecutive Drops: 1
Reset detected: No
Valid: Yes
EPU Errors: 6

PDI Errors: 0
Status Port 0: Has Link, Open, Has Comm
RX Error Port 0: 0
Forwarded RX Error Port 0: 6

Invalid Frame Port 0: 0
Lost Link Port 0: 0
Status Port 1: Has Link, Open, Has Comm
RX Error Port 1: 0
Forwarded RX Error Port 1: 13

Invalid Frame Port 1: 0
Lost Link Port 1: 0
}}}

Checksum error with replacement board on PRL.
{{{
Full name: /Motors/EtherCAT Device (r_forearm_roll_motor)
Component: EtherCAT Device (r_forearm_roll_motor)
Hardware ID: 68-05005-01860
Level: Error
Message: Checksum error on status data

Configuration: good
Name: r_forearm_roll_motor
Position: 30
Product code: WG05 (6805005) Firmware Revision 1.20, PCB Revision F.02
Robot: PR2
Motor: Maxon 310009
Serial Number: 68-05005-01860
Nominal Current Scale: 0.000500
Nominal Voltage Scale: 0.006250
HW Max Current: 3.250000
SW Max Current: 1.720000
Speed Constant: 178.000000
Resistance: 2.520000
Motor Torque Constant: 0.053800
Pulses Per Revolution: 1200
Encoder Reduction: -1.000000
Safety Disable Status: DISABLED, UNDERVOLTAGE (03)
Safety Disable Status Hold: DISABLED, UNDERVOLTAGE (03)
Safety Disable Count: 14
Undervoltage Count: 4
Over Current Count: 0
Board Over Temp Count: 0
Bridge Over Temp Count: 0
Operate Disable Count: 1
Watchdog Disable Count: 11
MBX Command IRQ Count: 74
PDI Timeout Error Count: 0
PDI Checksum Error Count: 0
Supply Current: 0.509128
Configured Offset A: -0.021000
Configured Offset B: 0.003000
Mailbox Write Errors: 0
Mailbox Read Errors: 0
Mailbox Retries: 0
Mailbox Retry Errors: 0
Watchdog Limit: 100ms
Mode: UNDERVOLTAGE, SAFETY_LOCKOUT
Digital out: 254
Programmed pwm value: 0
Programmed current: 0.000000
Measured current: -0.161000
Timestamp: 2657337456
Encoder count: -127408
Encoder index pos: -127408
Num encoder_errors: 0
Encoder status: 15
Calibration reading: 15
Last calibration rising edge: -104842
Last calibration falling edge: -50291
Board temperature: 38.562500
Max board temperature: 38.812500
Bridge temperature: 41.687500
Max bridge temperature: 44.187500
Supply voltage: 17.450001
Motor voltage: 0.043750
Current Loop Kp: 16
Current Loop Ki: 4
Motor Voltage Error %: 30.927817
Max Motor Voltage Error %: 57.016365
Filtered Voltage Error %: 31.147152
Max Filtered Voltage Error %: 43.619246
Current Error: -0.000064
Max Current Error: 0.058038
Filtered Current Error: 0.000015
Max Filtered Current Error: 0.013445
Motor Resistance Estimate: 6.595920
Packet count: 63128
Drops: 8
Consecutive Drops: 0
Max Consecutive Drops: 1
Reset detected: No
Valid: Yes
EPU Errors: 12
PDI Errors: 0
Status Port 0: Has Link, Open, Has Comm
RX Error Port 0: 0
Forwarded RX Error Port 0: 12
Invalid Frame Port 0: 0
Lost Link Port 0: 0
Status Port 1: Has Link, Open, Has Comm
RX Error Port 1: 0
Forwarded RX Error Port 1: 12
Invalid Frame Port 1: 0
Lost Link Port 1: 0
}}}

Error from board on PRH
{{{
Full name: /Motors/EtherCAT Device (r_forearm_roll_motor)
Component: EtherCAT Device (r_forearm_roll_motor)
Hardware ID: 68-05005-01552
Level: Error
Message: Checksum error on status data

Configuration: good
Name: r_forearm_roll_motor
Position: 30
Product code: WG05 (6805005) Firmware Revision 1.20, PCB Revision F.02
Robot: PR2
Motor: Maxon 310009
Serial Number: 68-05005-01552
Nominal Current Scale: 0.000500
Nominal Voltage Scale: 0.006250
HW Max Current: 3.250000
SW Max Current: 1.720000
Speed Constant: 178.000000
Resistance: 2.520000
Motor Torque Constant: 0.053800
Pulses Per Revolution: 1200
Encoder Reduction: -1.000000
Safety Disable Status: ENABLED (00)
Safety Disable Status Hold: ENABLED (00)
Safety Disable Count: 26
Undervoltage Count: 16
Over Current Count: 0
Board Over Temp Count: 0
Bridge Over Temp Count: 0
Operate Disable Count: 1
Watchdog Disable Count: 11
MBX Command IRQ Count: 210
PDI Timeout Error Count: 0
PDI Checksum Error Count: 0
Supply Current: 0.141264
Configured Offset A: -0.002000
Configured Offset B: -0.016500
Mailbox Write Errors: 0
Mailbox Read Errors: 0
Mailbox Retries: 0
Mailbox Retry Errors: 0
Watchdog Limit: 100ms
Mode: OFF
Digital out: 254
Programmed pwm value: 0
Programmed current: 0.000000
Measured current: 0.031500
Timestamp: 569883968
Encoder count: 423896
Encoder index pos: 423896
Num encoder_errors: 0
Encoder status: 15
Calibration reading: 15
Last calibration rising edge: 427198
Last calibration falling edge: 427234
Board temperature: 35.687500
Max board temperature: 38.062500
Bridge temperature: 38.000000
Max bridge temperature: 43.500000
Supply voltage: 70.956253
Motor voltage: -0.106250
Current Loop Kp: 16
Current Loop Ki: 4
Motor Voltage Error %: -4.269511
Max Motor Voltage Error %: 64.600460
Filtered Voltage Error %: 4.304521
Max Filtered Voltage Error %: 59.810880
Current Error: -0.000008
Max Current Error: 0.042598
Filtered Current Error: 0.000056
Max Filtered Current Error: 0.007211
Motor Resistance Estimate: 4.878233
Packet count: 3994
Drops: 0
Consecutive Drops: 0
Max Consecutive Drops: 0
Reset detected: No
Valid: Yes
EPU Errors: 6
PDI Errors: 0
Status Port 0: Has Link, Open, Has Comm
RX Error Port 0: 0
Forwarded RX Error Port 0: 6
Invalid Frame Port 0: 0
Lost Link Port 0: 0
Status Port 1: Has Link, Open, Has Comm
RX Error Port 1: 0
Forwarded RX Error Port 1: 6
Invalid Frame Port 1: 0
Lost Link Port 1: 0
}}}

**Comment by ahendrix** [comment]
[dking] I changed EML to break process data in to 20byte chunks and ran on a WG006.  The process data is now being broken across 8 different packets.  After 2+ hours of operation I've seen no checksum errors. 

Melonee mentioned that the checksum error occured more often when PRL was under heavy CPU load.   It could be possible to get corrupted status data if the operating system has transmitted the packets in a different order.  In this case, the SyncManager would reject the buffer read from second packet because the first packet would not have already accessed the first byte of the buffer.  

It should be possible to detect this situation by monitoring the working counters coming back from the process data packets.  Also, the EML library should also see the packets coming back in the incorrect order.  The '''EtherCAT Master''' diagnostics already includes counters for out-of-order receives with the '''RX Bad Order''' value.

I'll grab the '''EtherCAT Master''' diagnostics from the next checksum error.  It might also be possible to pull this information from the old diagnostics logs.

**Comment by ahendrix** [comment]
[dking] Disabling SyncManager on status data leads to checksum errors occurring with MTBF < 2 seconds.

{{{
--- src/wg0x.cpp    (revision 31507)
+++ src/wg0x.cpp    (working copy)
@@ -309,7 +309,7 @@
   (*pd)[0].ALEventEnable = true;

   (*pd)[1] = EC_SyncMan(STATUS_PHY_ADDR, base_status);
-  (*pd)[1].ChannelEnable = true;
-  (*pd)[1].ChannelEnable = false;
  
  (_pd)[2] = EC_SyncMan(MBX_COMMAND_PHY_ADDR, MBX_COMMAND_SIZE, EC_QUEUED, EC_WRITTEN_FROM_MASTER);
  (_pd)[2].ChannelEnable = true;
  }}}

**Comment by ahendrix** [comment]
[dking] Another error (from last week).   Looks like packets are getting dropped because the OS is delivering them late (likely due to heavy CPU load).   After reviewing the code, I found that ethercat_hardware is not re-trying after a packet drop.  Because of this some of the boards would be run with new data, some with old data, and one board with half-old or half-new data.   I've checked in a fix into trunk with r31538.

{{{
Full name: /Motors/EtherCAT Master
Component: EtherCAT Master
Hardware ID: 
Level: Error
Message: Motors halted
Motors halted: true
EtherCAT devices (expected): 37
EtherCAT devices (current): 37
Interface: ecat0
Reset state: 0
Average roundtrip time (us): 371.0599
Maximum roundtrip time (us): 20086.0000
EtherCAT Process Data txandrx errors: 11
Input Thread: Running
Sent Packets: 20223862
Received Packets: 20223842
Collected Packets: 20223841
Dropped Packets: 21
TX Errors: 0
TX Network Down: 0
TX Queue Full: 0
RX Runt Packet: 0
RX Not EtherCAT: 0
RX Other EML: 0
RX Bad Index: 0
RX Bad Sequence: 0
RX Duplicate Sequence: 0
RX Duplicate Packet: 0
RX Bad Order: 0
RX Late Packet: 20
RX Late Packet RTT: 21460
RX Late Packet Avg RTT: 27383.650000
}}}

{{{
Full name: /Motors/EtherCAT Device (r_forearm_roll_motor)
Component: EtherCAT Device (r_forearm_roll_motor)
Hardware ID: 68-05005-01549
Level: Error
Message: Checksum error on status data
Configuration: good
Name: r_forearm_roll_motor
Position: 30
Product code: WG05 (6805005) Firmware Revision 1.20, PCB Revision F.02
Robot: PR2
Motor: Maxon 310009
Serial Number: 68-05005-01549
Nominal Current Scale: 0.000500
Nominal Voltage Scale: 0.006250
HW Max Current: 3.250000
SW Max Current: 1.720000
Speed Constant: 178.000000
Resistance: 2.520000
Motor Torque Constant: 0.053800
Pulses Per Revolution: 1200
Encoder Reduction: -1.000000
Safety Disable Status: ENABLED (00)
Safety Disable Status Hold: ENABLED (00)
Safety Disable Count: 35
Undervoltage Count: 14
Over Current Count: 0
Board Over Temp Count: 0
Bridge Over Temp Count: 0
Operate Disable Count: 1
Watchdog Disable Count: 22
MBX Command IRQ Count: 80
PDI Timeout Error Count: 0
PDI Checksum Error Count: 0
Supply Current: 0.141264
Configured Offset A: -0.001500
Configured Offset B: 0.011000
Mailbox Write Errors: 0
Mailbox Read Errors: 0
Mailbox Retries: 0
Mailbox Retry Errors: 0
Watchdog Limit: 100ms
Mode: OFF
Digital out: 254
Programmed pwm value: 0
Programmed current: 0.000000
Measured current: -0.026000
Timestamp: 1299297352
Encoder count: 432640
Encoder index pos: 432640
Num encoder_errors: 0
Encoder status: 12
Calibration reading: 15
Last calibration rising edge: 432903
Last calibration falling edge: 432967
Board temperature: 37.562500
Max board temperature: 40.312500
Bridge temperature: 40.000000
Max bridge temperature: 45.875000
Supply voltage: 70.968750
Motor voltage: -0.100000
Current Loop Kp: 16
Current Loop Ki: 4
Motor Voltage Error %: 16.125427
Max Motor Voltage Error %: 107.035585
Filtered Voltage Error %: 16.184484
Max Filtered Voltage Error %: 99.430730
Current Error: 0.000000
Max Current Error: 0.082707
Filtered Current Error: 0.000014
Max Filtered Current Error: 0.014975
Motor Resistance Estimate: 6.671830
Packet count: 10248
Drops: 10
Consecutive Drops: 0
Max Consecutive Drops: 2
Reset detected: No
Valid: Yes
EPU Errors: 0
PDI Errors: 0
Status Port 0: Has Link, Open, Has Comm
RX Error Port 0: 0
Forwarded RX Error Port 0: 0
Invalid Frame Port 0: 0
Lost Link Port 0: 0
Status Port 1: Has Link, Open, Has Comm
RX Error Port 1: 0
Forwarded RX Error Port 1: 0
Invalid Frame Port 1: 0
Lost Link Port 1: 0
}}}
